packages/ws-server/src/webSocketServer.ts

```diff
@@ -46,11 +46,20 @@ const LOGGER_PREFIX = 'WebSocket:';
 
 const CHAIN_ID = relay.eth().chainId();
 const DEFAULT_ERROR = predefined.INTERNAL_ERROR();
+const MAX_CONNECTIONS = parseInt(process.env.CONNECTION_LIMIT || '10');
+let connectedClients = 0;
 
 app.ws.use((ctx) => {
     ctx.websocket.id = relay.subs()?.generateId();
     logger.info(`${LOGGER_PREFIX} new connection ${ctx.websocket.id}`);
 
+    if (connectedClients >= MAX_CONNECTIONS) {
+        ctx.websocket.send("Connections exceeded");
+        ctx.websocket.close();
+        return;
+    }
+    connectedClients++;
```

Comment:
I'm curious if the ordering here is correct. When the 11th client connects, the connection would be closed and the `close` event will be triggered, which will decrement `connectedClients`. I think then `connectedClients` will become `9` instead of remaining unchanged.

Could you please add a test to verify this?

Reply:
It's surprisingly hard to find documentation on whether the `onclose` event (or any other websocket event) is fired before `close()` returns or if it can be fired sometime later.  Is this something that's known for websocket `close()` in particular, or is a general behavior of JavaScript that I don't know?

If the answer is: "can be fired sometime after the return of `close()`" then a test with the existing code may sometimes work and sometimes not due to a race condition.  Maybe?

Reply:
@david-bakin-sl, does this helps answer your question? I think is fired **after always**.

https://github.com/websockets/ws/blob/7.4.6/lib/websocket-server.js#L146

Reply:
Some tests were added to verify the connection limits.  Some refactoring of the implementation was also done.